sklearn/decomposition/_pca.py

```diff
@@ -482,9 +488,12 @@ def _fit(self, X):
                 "PCA does not support sparse input. See "
                 "TruncatedSVD for a possible alternative."
             )
+        # Raise an error for torch input and arpack or randomized solver.
+        if self.svd_solver in ["arpack", "randomized"] and _is_torch_namespace(xp):
+            raise TypeError(self._pca_torch_arpack_solver_error_message)
```

Comment:
For `randomized_svd`, apparently `xp.linalg.lu` is not standardized (yet?):

- https://data-apis.org/array-api/latest/extensions/linear_algebra_functions.html

so we might want to implement a namespace specific compatibility stub but we might to defer that to a follow-up PR.

Reply:
I have opened https://github.com/data-apis/array-api/issues/627 to discuss the possibility to standardize an API for LU factorization.

Reply:
> I would rather raise a `ValueError` stating that `svd_solver="arpack"` is only valid when the input data is passed as a numpy array and avoid special casing PyTorch: ideally all Array API compliant namespaces should work, not just PyTorch.

👍  for this. I think having a difference between "works with Numpy" and "works with Array API support enabled" is fine. In the case where there is no Array API support I'd even go as far as making it so that you receive an error if you pass a Numpy array when Array API support is enabled (even though it works if you disable Array API support).

Having Array API support but only for some libraries (say PyTorch but not Cupy) will be confusing and surprising to users.

At the same time I think it is fine to special case things for performance reasons. This means we have one path that works for all Array API compliant inputs and in addition a special PyTorch path that is significantly faster. The important thing is that there is a implementation that works with all Array API compliant inputs/namespaces.

Reply:
I changed the default normalizer to "qr", as this will allow to use `torch.linalg.rq` and `np.linalg.qr` without much overhead with `xp.linalg.qr`. Also changed `TypeError` to `ValueError`.